pkg/client/client.go

```diff
@@ -0,0 +1,135 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const LastAppliedConfigurationAnnotationKey = "toolchain.dev.openshift.com/last-applied-configuration"
+
+var log = logf.Log.WithName("apply_client")
+
+type ApplyClient struct {
+	cl     client.Client
+	scheme *runtime.Scheme
+}
+
+// NewApplyClient returns a new ApplyClient
+func NewApplyClient(cl client.Client, scheme *runtime.Scheme) *ApplyClient {
+	return &ApplyClient{cl: cl, scheme: scheme}
+}
+
+// CreateOrUpdateObject creates the object if is missing and if the owner object is provided, then it's set as a controller reference.
+// If the objects exists then when the spec content has changed (based on the content of the annotation in the original object) then it
+// is automatically updated. If it looks to be same then based on the value of forceUpdate param it updates the object or not.
+// The return boolean says if the object was either created or updated
+func (p ApplyClient) CreateOrUpdateObject(obj runtime.Object, forceUpdate bool, owner v1.Object) (bool, error) {
```

Comment:
do we really need to wrap the call to `createOrUpdateObj` in `CreateOrUpdateObject`?

Reply:
any other idea?